External/SPMmodified/xASL/xASL_im_FillNaNs.m

```diff
@@ -191,62 +254,162 @@
+                % of the current column
+                [diffIs, minInd, maxInd, thresholdIs] = xASL_im_Extrapolate_FixEdges(imThisCol, 1);
+
+                % Now we check if we should remove edge voxels
+                % We try this nEdgeVoxels times (default=5)
+                nLayersProcessed = 0;
+                while nLayersProcessed<nEdgeVoxels
+                    % 
+                    if diffIs(end)>thresholdIs
+                        imThisCol(maxInd) = NaN; % remove the edge voxel
+                        % If we have changed a first or last non-nan voxel, we rerun the difference determination 
+                        [diffIs, minInd, maxInd, thresholdIs] = xASL_im_Extrapolate_FixEdges(imThisCol, 1);
+                    end
+                    nLayersProcessed = nLayersProcessed+1;
+                end
 
-            % Put back together to the original image
+                meanDiff = xASL_stat_MeanNan(diffIs);
+
+                % Create a robust last value, from 6 voxels
+                % This avoids extrapolating an extreme value at the edge
+                voxels6 = imThisCol(maxInd-5:maxInd); % select the last 6 voxels
+                voxels6 = voxels6(isfinite(voxels6)); % select those without NaNs
+                robustMedian = median(voxels6);
+                robustDiff = median(voxels6(2:end) - voxels6(1:end-1)); % mean difference
+                numVoxels = length(voxels6)/2;
+                robustValue = robustMedian + numVoxels*robustDiff;
+                % Now take the mean of the actual last value and the robust one
+                robustValue = mean([robustValue imThisCol(maxInd)]);
+
+                % Apply this
+                imThisCol(maxInd+1:end) = robustValue + (1:nCol-maxInd) * meanDiff;
+                % Put this column back in the image
+                imCol(:, ColumnIs) = imThisCol;
+            end
+    
+            %% 4. Reshape 2D back to 3D
             switch(DimensionIs)
                 case 1
                     IM = reshape(imCol, [nX,nY,nZ]);
                 case 2
-					% For 2D matrix and 3D matrix, the reshaping back is done differently
-					if ndims(IM)<3
-						IM = shiftdim(reshape(imCol,[nY,nX]),1);
-					else
-						IM = shiftdim(reshape(imCol,[nY,nZ,nX]),2);
-					end
+				    % For 2D matrix and 3D matrix, the reshaping back is done differently
+				    if ndims(IM)<3
+					    IM = shiftdim(reshape(imCol,[nY,nX]),1);
+				    else
+					    IM = shiftdim(reshape(imCol,[nY,nZ,nX]),2);
+				    end
                 case 3
                     IM = reshape(imCol', [nX,nY,nZ]);
             end % switch
         end % for DimensionIs
     end % sum(isnan(IM))
-end % for Iteration
+end
 
 % Any remaining NaNs had to be inside, so we interpolate them (setting them
 % to zeros creates artifacts!)
 IM = xASL_im_ExtrapolateSmoothOverNaNs(IM, 0);
 
+end
+
+
+function [diffIs, minInd, maxInd, thresholdIs] = xASL_im_Extrapolate_FixEdges(imThisCol, bSide)
```

Comment:
I miss the rationale for this statistics and checking. We should simply replace nans by linearly extrapolated values and not calculate stats etc. 

I understand that there were some errors, but this entire construction is extremely complicated.

Reply:
See whatsapp. I will add a rationale, but this part didn't work well, the distance-based outlier exclusion worked better.

Reply:
OK. I'll check in detail.